src/controller/CHIPDeviceController.cpp

```diff
@@ -649,74 +574,114 @@ void ChipDeviceController::OnRendezvousError(CHIP_ERROR err)
 
     if (mPairingDelegate != nullptr)
     {
-        mPairingDelegate->OnPairingComplete(err);
+        mPairingDelegate->OnPairingComplete(status);
+    }
+
+    if (mDeviceBeingPaired != kNumMaxActiveDevices)
+    {
+        DeviceController::ReleaseDevice(mDeviceBeingPaired);
+        mDeviceBeingPaired = kNumMaxActiveDevices;
     }
 }
 
-void ChipDeviceController::OnRendezvousComplete()
+void DeviceCommissioner::OnRendezvousError(CHIP_ERROR err)
 {
-    OnRendezvousError(CHIP_NO_ERROR);
+    RendezvousCleanup(err);
 }
 
-void ChipDeviceController::OnRendezvousStatusUpdate(RendezvousSessionDelegate::Status status, CHIP_ERROR err)
+void DeviceCommissioner::OnRendezvousComplete()
 {
-    if (mOnError != nullptr && err != CHIP_NO_ERROR)
+    CHIP_ERROR err  = CHIP_NO_ERROR;
+    Device * device = nullptr;
+    VerifyOrExit(mDeviceBeingPaired < kNumMaxActiveDevices, err = CHIP_ERROR_INVALID_DEVICE_DESCRIPTOR);
+    device = &mActiveDevices[mDeviceBeingPaired];
+    mPairedDevices.Insert(device->GetDeviceId());
+    mPairedDevicesUpdated = true;
+
+    if (mStorageDelegate != nullptr)
+    {
+        SerializedDevice serialized;
+        device->Serialize(serialized);
+        PERSISTENT_KEY_OP(device->GetDeviceId(), kPairedDeviceKeyPrefix, key,
+                          mStorageDelegate->SetKeyValue(key, Uint8::to_const_char(serialized.inner)));
+    }
+
+exit:
+    if (err == CHIP_NO_ERROR)
+    {
+        RendezvousCleanup(CHIP_NO_ERROR);
+    }
+    else
+    {
+        OnRendezvousError(err);
+    }
+}
+
+void DeviceCommissioner::OnRendezvousStatusUpdate(RendezvousSessionDelegate::Status status, CHIP_ERROR err)
+{
+    Device * device = nullptr;
+    if (mDeviceBeingPaired >= kNumMaxActiveDevices)
     {
-        mOnError(this, mAppReqState, err, nullptr);
+        ExitNow();
     }
 
+    device = &mActiveDevices[mDeviceBeingPaired];
     switch (status)
     {
     case RendezvousSessionDelegate::SecurePairingSuccess:
-        ChipLogProgress(Controller, "Remote device completed SPAKE2+ handshake\n");
-        mPairingSession       = mRendezvousSession->GetPairingSession();
-        mSecurePairingSession = &mPairingSession;
-
-        if (mOnNewConnection)
-        {
-            ChipLogProgress(Controller, "Will Call on mOnNewConnection(%p)\n", mOnNewConnection);
-            mOnNewConnection(this, nullptr, mAppReqState);
-        }
+        ChipLogDetail(Controller, "Remote device completed SPAKE2+ handshake\n");
+        mRendezvousSession->GetPairingSession().Serializable(device->GetPairing());
```

Comment:
The naming here is a bit weird.  Should that function be called `Serialize`?

Reply:
It's actually outputting the `Serializable` object, instead of serializing the data. Would `ToSerializable` be a better name?

Reply:
Yes, `ToSerializable` would be better.